src/test/unittest/ctx_filter.py

```diff
@@ -0,0 +1,156 @@
+"""
+
+import sys
+import itertools
+
+import futils
+import requirements
+import context as ctx
+import valgrind as vg
+
+
+class CtxFilter:
+    """
+    Generates contexts based on provided
+    configuration and test requirements
+    """
+
+    def __init__(self, config, tc):
+        self.reqs = requirements.get_requirements(tc)
+        self.tc = tc
+        self.config = config
+        self.msg = futils.Message(config.unittest_log_level)
+
+    def filter_contexts(self, config_ctx, test_ctx):
+        """
+        Return contexts that should be used in execution based on
+        contexts provided by config and test case
+        """
+        if not test_ctx:
+            return [c for c in config_ctx if not c.explicit]
+        return [c for c in config_ctx if c in test_ctx]
+
+    def get_contexts(self):
+        """
+        Generate a list of context based on configuration and
+        test requirements
+        """
+        params = self.get_configurational_params()
+
+        ctx_params = itertools.product(*params.values())
+        ctxs = []
+        for cp in ctx_params:
+            kwargs = dict(zip(params.keys(), cp))
+            c = ctx.Context(**kwargs)
+            c.cwd = self.tc.cwd
+            ctxs.append(c)
+        return ctxs
+
+    def get_configurational_params(self):
+        """
+        Get special test parameters, like build, file system or valgrind tool
+        which final content depends on test requirements, user configuration
+        and/or provided test plan.
+        """
+        params = {}
+
+        params['build'] = self.filter_build()
+        params['fs'] = self.filter_fs()
+        if sys.platform != 'win32':
+            params['valgrind'] = self.filter_valgrind()
+        return params
+
+    def filter_build(self):
+        """
+        Acquire build type for the test to be run based on configuration
+        and test requirements
+        """
+        kwargs = self.reqs.build_kwargs
+        builds = []
+        for b in self.filter_contexts(self.config.build, self.reqs.build):
+            try:
+                builds.append(b(**kwargs))
+            except futils.Skip as s:
+                self.msg.print_verbose('{}: SKIP: {}'.format(self.tc, s))
+
+        return builds
+
+    def filter_fs(self):
+        """
+        Acquire file system for the test to be run based on configuration
+        and test requirements
+        """
+        kwargs = self.reqs.fs_kwargs
+        kwargs['tc_dirname'] = self.tc.tc_dirname
+
+        if self.reqs.fs == ctx.Non:
+            return [ctx.Non(**kwargs), ]
+        else:
+            fss = []
+            for f in self.filter_contexts(self.config.fs, self.reqs.fs):
+                try:
+                    fss.append(f(**kwargs))
+                except futils.Skip as s:
+                    self.msg.print_verbose('{}: SKIP: {}'.format(self.tc, s))
+
+            return fss
+
+    def filter_valgrind(self):
```

Comment:
Would it be possible to group filters/requirements, and in general code, not by it's function (e.g., filter) in the framework, but by its type (e.g., valgrind, fs)?
I'm asking because I'd prefer to have the code that sets, parses, filters the same thing in the same location.